pkg/cloud-controller-manager/loadbalancer.go

```diff
@@ -0,0 +1,396 @@
+	}
+	// check if the port of the secondary service overlaps with the primary service and other secondary services
+	if err := l.checkPortOverlap(primary, secondary); err != nil {
+		return nil, fmt.Errorf("check port overlap failed, primary service: %s/%s, secondary service: %s/%s, error: %w",
+			primary.Namespace, primary.Name, secondary.Namespace, secondary.Name, err)
+	}
+	// delete the original load balancer if existing
+	if err := l.deleteLoadBalancer(clusterName, secondary); err != nil {
+		return nil, err
+	}
+	// update secondary service load balancer IP
+	return &secondary.Status.LoadBalancer, l.updateSecondaryServiceLoadBalancerIP(primary.Status.LoadBalancer.Ingress[0].IP, primary, secondary)
+}
+
+func (l *LoadBalancerManager) UpdateLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) error {
+	if _, err := l.EnsureLoadBalancer(ctx, clusterName, service, nodes); err != nil {
+		return fmt.Errorf("update load balancer failed, error: %w", err)
+	}
+	return nil
+}
+
+func (l *LoadBalancerManager) EnsureLoadBalancerDeleted(ctx context.Context, clusterName string, service *v1.Service) error {
+	primarySvc, err := l.getPrimaryService(service)
+	if err != nil {
+		return err
+	}
+	// do nothing for the secondary service
+	if primarySvc != nil && primarySvc != service {
+		return nil
+	}
+
+	return l.deleteLoadBalancer(clusterName, service)
+}
+
+func (l *LoadBalancerManager) createOrUpdateLoadBalancer(name, clusterName string, service *v1.Service) error {
+	lb, err := l.lbClient.Get(l.namespace, name, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	newLB := l.constructLB(lb, service, name, clusterName)
+	if errors.IsNotFound(err) {
+		_, err = l.lbClient.Create(newLB)
+	} else {
+		_, err = l.lbClient.Update(newLB)
+	}
+
+	return err
+}
+
+func (l *LoadBalancerManager) constructLB(oldLB *lbv1.LoadBalancer, service *v1.Service, name, clusterName string) *lbv1.LoadBalancer {
+	var lb *lbv1.LoadBalancer
+
+	// If the error returned by Get Interface is ErrNotFound, the returned lb would not be nil, but the name of the lb is empty.
+	if oldLB == nil || oldLB.Name == "" {
+		lb = &lbv1.LoadBalancer{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: lbv1.SchemeGroupVersion.String(),
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: l.namespace,
+				Name:      name,
+			},
+		}
+	} else {
+		lb = oldLB.DeepCopy()
+	}
+
+	if lb.Annotations == nil {
+		lb.Annotations = make(map[string]string)
+	}
+	lb.Annotations[pkgctllb.AnnotationKeyNetwork] = service.Annotations[KeyNetwork]
+	lb.Annotations[pkgctllb.AnnotationKeyProject] = service.Annotations[KeyProject]
+	lb.Annotations[pkgctllb.AnnotationKeyNamespace] = service.Annotations[KeyNamespace]
+	lb.Annotations[pkgctllb.AnnotationKeyCluster] = clusterName
+
+	if lb.Labels == nil {
+		lb.Labels = make(map[string]string)
+	}
+	lb.Labels[clusterNameKey] = clusterName
```

Comment:
the value of `clusterName` seems always be `kubernetes`?

Reply:
```
      --cluster-name string                                                                                                                                         
                The instance prefix for the cluster. (default "kubernetes")

```